[CR] Smart engine controller and hybrid car

Aivean opened the pull request:
#### Summary

SUMMARY: Features "Added smart engine controller vehicle part and Hybrid Car"

#### Additional summary

[Youtube demo 1, HUMVEE (turn on subtitles!)](https://www.youtube.com/watch?v=wuTSkvPsE3U)
[Youtube demo 2, Hybrid car (turn on subtitles!)](https://youtu.be/vzoaH-4IcL8)

"Smart controller" is a vehicle part, that, when installed in a multi-engine vehicle and active, automatically switches combustion and electric engines on and off, minimizing fuel consumption and optimizing power output and battery charge rate.

Simplified ruleset: 
1. When throttling, maximize acceleration
2. Keep battery at 90%
3. Minimize fuel consumption

"Hybrid car" is a car with additional electric motor, two medium storage batteries and a smart engine controller.

#### Purpose of change

Originally I was trying to implement real-world hybrid electric vehicles (think Toyota Prius), that have two engines, electric and gas and switch between them semi-intelligently based on different factors, such as speed, requested acceleration and battery state. 

But after a few iterations trying different hand-crafted rules I realized that this feature can be more generic, allowing to improve usability of any vehicle with multiple engines.


#### Describe the solution

Changes:
* vehicle part ["smart engine controller"](https://github.com/CleverRaven/Cataclysm-DDA/compare/master...Aivean:vehicle-hybrid-controller?expand=1#diff-261c00b672330bfca5c46b62b2799f57R1877) is added   
* new vehicle type "Hybrid Car" is added
   * is similar to "Car", except has two medium storage batteries, additional electric engine and "smart engine controller"
   * spawn groups and rates are similar to that of an "electric car"
* some minor method extraction/refactoring was performed when I needed to reuse the parts of existing code
   * [several new methods were introduced into `vehicle`](https://github.com/CleverRaven/Cataclysm-DDA/compare/master...Aivean:vehicle-hybrid-controller?expand=1#diff-908c99735d088739104b08db8d9cd56cR1659)
   * ids of the "battery" vehicle parts are added to part cache (to speed up battery level calculation)
   * `has_enabled_smart_controller` bool flag is added to the vehicle cache   
   * `avg_velocity`  (exponential average velocity of the vehicle) and `smart_controller_state` (smart controller state cache) were added to the vehicle state (transient, not serialized).
    
For smart controller, a single method is added ([`vehicle::smart_controller_handle_turn`](https://github.com/CleverRaven/Cataclysm-DDA/compare/master...Aivean:vehicle-hybrid-controller?expand=1#diff-eb2a47f3a4b3b664db5a33a8e20fbcb0R120
)) that is called from `vehicle::thrust` and `vehicle::idle`. It may be called two times per turn, but will be executed only once (due to caching).

**NOTE**: no changes to current vehicle math/mechanics were made. Smart controller entirely relies on current math to make its decisions.

**Logic summary:** 

If there is an active smart controller and vehicle has two or more controllable engines (controllable engines are currently gas and electric engines) all combinations (on/off states) of controllable engines are tried and for each combinations following factors are considered:
* battery charge %
    * when charge < 25%, electric motors won't turn on
    * when charge < 90%, configuration with increased charge rate will be prioritized
* acceleration  vs requested acceleration
    * when user requests acceleration, configurations that have sufficient (or maximal) acceleration are prioritized
* safe velocity vs requested velocity
   * configurations that are safe for engines take priority
* fuel usage
   * fuel consumption (measured in energy, J) is minimized
* battery charge rate

If no configuration better than current one is found, current configuration remains. 

#### Describe alternatives you've considered

Initially I tried several variants of fixed hand-picked rules, in the vein of: "when battery level is < 90%, turn off electric motor, turn on gas motor;   when load > 50%,  turn on all engines".  Unfortunately, in practice, no such rules can be good enough. Current engine math allows for tremendous amount of corner cases. To name a few:
* two engines that have significant disparity in power working together are less efficient than only one engine due to "multiple engines overhead coefficient" (what about three or more engines?).  On the other hand, engines with similar power output suffer less from that effect and, for some cases, running both engines simultaneously can actually be better than running just one of them.
* depending on the current speed (and, thus, engine load),  larger engine can be more efficient than smaller one (and vice versa); that applies to both gas and electric engines

At some point I just gave up on trying to manually address all corner cases (I kept stumbling into new ones) and decided to implement a generic solution that, as a side benefit, will continue to work even if engine math is changed in the future.

#### Testing

I playtested following cases:

* new hybrid car
* light vehicle (mini-car) with two engines
* heavy vehicles (humvee and APC) with two and three engines of different size and type:
    *  v12 and v6 diesel 
    * 1 cyl gas,  v6 diesel and electric (regular and enhanced)
    * v6 diesel and electric
* boat with two engines (gas and electric)
* electric scooter with two engines (gas and electric)
* electric bike with three engines (muscle engine is not controlled)

Also I tried:
* driving on and off road to ensure that there is no oscillation between engine states
* overload engines to check behavior when engine dies mid-drive
* running out of energy/gas
* engines with faults that won't start


#### Additional context


**Performance**

If engine is off or smart controller is off, nothing is computed.  ["Smart controller on"](https://github.com/CleverRaven/Cataclysm-DDA/compare/master...Aivean:vehicle-hybrid-controller?expand=1#diff-3d6b9eaf8fd915cd10858174d45c8d6aR5561) is a boolean flag and it is cached.

Although "all engine state combinations" is O(2^n), `n` here is limited to 5.  (Smart controller doesn't support more than 5 engines), so worst case number of combinations checked would be 31  ("all off" state is not checked). In average game engines number is rarely bigger than 2 or 3.

For each considered engine combination, following things are computed:
* battery level and battery capacity
   * one loop through cached battery parts
* wheel traction
   * one loop through cached wheels
* battery charge rate
  *  loop through cached engines, alternators, epowered accessories, solar panels, wind turbines
* `current_acceleration`  (cached)
* `safe_ground_velocity` 
   * loop cached engines
   * cached coefficients
* engine fuel and epower usage
   * loop through cached engines

Every loop above goes through pre-cached part indices, which typically have "less than 10" items.

In average case, factors above (they determine engine configuration) are stored in ["smart controller state cache"](https://github.com/CleverRaven/Cataclysm-DDA/compare/master...Aivean:vehicle-hybrid-controller?expand=1#diff-908c99735d088739104b08db8d9cd56cR104), so they get evaluated only one time per turn (for current configuration), and, if no significant change is detected, no further optimization is performed.

When there is a significant change (often happens when user changes speed, or vehicle rides over something), all 2^n configurations are checked, evaluating factors above per each engine state combination and finding best one.

For stationary/cruising vehicles with turned on engine and smart controller full evaluation will happen fairly rare (usually when battery level changes for several %).


**Regarding fuel consumption**

At some point I realized that within current CDDA engine mechanics, it's not possible to save fuel in the same way as it's possible with hybrid vehicles IRL.  In real world, despite inefficiencies of energy conversion chain 

     chemical energy -> gas engine -> kinetic energy -> generator -> electric energy -> battery -> electric motor -> kinetic energy 

it's still better than simply

     chemical energy -> gas engine -> kinetic energy 

because in latter case gas engine works in suboptimal conditions and has way more energy loss than in former case.

In CDDA gas engines always operate with the same constant efficiency, so the former chain in CDDA is actually slightly worse than the latter one (precisely, it has additional 10% inefficiency of alternator and 10% inefficiency of electric motor).

**However**, when using smart controller with gas and electric motors, battery charge stays around 90%, which means, that when gas engine is on, alternator actually does useful work. Without smart controller, battery stays at 100% and alternator just wastes fuel.  Compared to single-gas-engine cars this can yield some marginal fuel savings.

Also, multi-engine car with smart controller has better acceleration than single-engine car (due to immediate additional engine engagement under load).

For multi-engine cars (with any engine types) smart controller benefit is obvious:  it retains good acceleration, at the same time it significantly reduces fuel usage in cruising situations (it leaves a single optimally chosen engine on).


**Debouncing**

There are several things that prevent fast back-and-forth oscillating between two engine configurations:

* there is a 15 in-game seconds timeout for shutting down gas engines after gas engine was started
     * it doesn't prevent starting additional engines (i.e. when there is sudden acceleration)
     * but it effectively prevents start-stop loop that involves gas engines
     * purely electric configurations can still oscillate fast, but that is intentional (it's ok to instantly power on/off electric engines IRL)
* instead of instant vehicle velocity, moving average velocity is used
* previous values of key factors that were used to find current engine configuration are cached and cache key is intentionally "fuzzy". Small changes to key factors (i.e., `load` change is less than 10%, or velocity difference is less than 1 mph) do not trigger smart controller optimization.



**Questions for community**

* currently "smart controller"  exists only as a vehicle part,  it's installed from "processor board", similar to "security system". Should I introduce a separate craftable item? If yes, what is the best recipe and requirements?
* currently it's possible to install multiple smart controllers, but only a single active one is enough to trigger the logic. I was also considering having smart controllers installed directly on the engines that are supposed to be controlled.  What do you think? Alternative approach has pros and cons:
   * ➖ it's a bit harder to implement than current one, specifically harder to limit overall number of installed controllers or otherwise limit the number of controlled engines
   * ➖ UX is more complicated: how to toggle all hybrid controllers? allow them to be toggled one by one like the engines currently?
   * ➕ more robust: gives player more control over what's happening with the individual engines

mlangsdorf commented:
If you got this to work then you're a mad genius and thank you very much for contributing.

mlangsdorf commented on src/vehicle_move.cpp:
```diff
@@ -116,6 +117,261 @@ int vehicle::slowdown( int at_velocity ) const
+    bool discharge_forbidden_soft = battery_level_percent <= 90;
+    bool discharge_forbidden_hard = battery_level_percent <= 25;
+    int target_charging_rate = ( max_battery_level == 0 || !discharge_forbidden_soft ) ? 0 :
+                               ( max_battery_level * 9 / 10 - cur_battery_level ) * 10 / ( 6 * 3 );
```
( 9 * max_battery_level - 10 * cur_battery_level ) / ( 6 * 3 )
saves one unnecessary division and shouldn't be any less clear.

Aivean commented on src/vehicle.cpp:
```diff
@@ -1153,19 +1158,32 @@ bool vehicle::is_part_on( const int p ) const
+           !eng.faults().count( fault_engine_belt_drive );
+}
 
-    return std::any_of( engines.begin(), engines.end(), [this, &alt]( int idx ) {
```
Good catch! I probably needed it in a non-final code version. I'll revert `is_alternator_on` method to its original state.

Aivean commented on src/vehicle.h:
```diff
@@ -1834,6 +1868,8 @@ class vehicle
         point pos;
         // vehicle current velocity, mph * 100
         int velocity = 0;
+        // vehicle exponential average velocity (av = (av + velocity) / 2), mph * 100
```
I don't think that it's wrong, but it may be confusing nevertheless:
https://en.wikipedia.org/wiki/Moving_average#Exponential_moving_average

![image](https://user-images.githubusercontent.com/2865203/86419884-89672080-bc89-11ea-9b01-a3be7071b53a.png)

I used a simplified version where α = 0.5:

`v_new =  v_prev + 0.5(v_cur - v_prev) =  0.5 v_cur + 0.5 v_prev`

What can I do to make it less confusing? Add a wikipedia link to the formula?

mlangsdorf replied:
just call it a continuous moving average, people will get what you mean.

Aivean commented on src/vehicle_move.cpp:
```diff
@@ -149,6 +405,10 @@ void vehicle::thrust( int thd, int z )
+    }
+
     // TODO: Pass this as an argument to avoid recalculating
     float traction = k_traction( g->m.vehicle_wheel_traction( *this ) );
```
I assume you're referring to `k_traction`.  This is actually a good idea, as it doesn't change after potential engine state change. Will do, thank you!

BevapDin commented on src/vehicle_move.cpp:
```diff
@@ -116,6 +117,261 @@ int vehicle::slowdown( int at_velocity ) const
+        bool gas_engine_to_shut_down = false;
+        for( size_t i = 0; i < c_engines.size(); ++i ) {
+            bool old_state = static_cast<bool>( ( prev_mask & ( 1 << i ) ) != 0 );
+            bool new_state = static_cast<bool>( ( mask & ( 1 << i ) ) != 0 );
```
```suggestion
            const bool new_state = ( mask & ( 1 << i ) ) != 0;
```
Same on the line above.

nexusmrsep commented:
It doesn't have to be here and now, but consider adding a pamphlet, a news article, snippets/tips etc. to hint players of it's functionality and benefits.

Aivean commented:
@mlangsdorf , regarding the performance, it seems more or less ok as it is. I did some basic profiling, and currently actively used smart controller (in a car that is constantly accelerating/braking to bust cache) takes about 1/4 of of the `thrust` method time on average:
![image](https://user-images.githubusercontent.com/2865203/86446278-95be9e00-bcc8-11ea-8b4e-6aba90fce6e8.png)
Note, this is total method time for 30s profiling session. Smart controller handler was not called every turn due to caching, even despite my best efforts to bust cache.

This time becomes essentially zero when vehicle is not actively driven.

However,  performance can certainly be improved. For instance, I discovered an "all parts" loop here:
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/src/vehicle.cpp#L3221

which takes at least half of the smart controller time.

I'll do more profiling and performance optimization and address all other issues tomorrow. 

Thanks for the code review.

nexusmrsep commented:
Idea from watching your videos (big thumbs up) - perhaps (as a future feature) smart controller could be instructed by the driver to cut off electrical engines at given battery capacity (ex. 20%) and re-enable them after battery is restored to another given capacity (ex. 95%). This would allow prolonged support from electric engines while reserving some battery for gas/diesel engines to start.

Aivean commented:
> Idea from watching your videos (big thumbs up) - perhaps (as a future feature) smart controller could be instructed by the driver to cut off electrical engines at given battery capacity (ex. 20%) and re-enable them after battery is restored to another given capacity (ex. 95%). This would allow prolonged support from electric engines while reserving some battery for gas/diesel engines to start.

That is what smart controller is currently doing (among other things).  There are two thresholds (currently hardcoded for simplicity).  
* I battery is < 10%, electric engines won't automatically engage even if user is accelerating.
* If battery is <=90%,  electric engines will engage only if there is a demand for acceleration or if battery charging rate is high
* if battery is >90%, discharge rate is not considered as a factor, so electric engines usually take precedence (as they are typically more fuel efficient than gas)

nexusmrsep commented:
I'm aware of that. What I'm trying to say here is that those numbers cannot be currently adjusted by the vehicle user and, considering complex player built deathmobiles, this might be what some players would want for their not-out-of-the-box custom machines.

Aivean commented:
> I'm aware of that. What I'm trying to say here is that those numbers cannot be currently adjusted by the vehicle user and, considering complex player built deathmobiles, this might be what some players would want for their not-out-of-the-box custom machines.

Gotcha. You're right, that might be a useful feature to have. I'll think about the best way to add this after the core functionality is merged.

Aivean commented:
I've addressed all code issues from reviews. 

In addition, I did following improvements:
* for stationary vehicles (when velocity and acceleration demand is zero) [no velocity-related computations are performed in smart controller handler](https://github.com/CleverRaven/Cataclysm-DDA/pull/41780/commits/3d89fa6467382f7db524097527b7871372e22915#diff-eb2a47f3a4b3b664db5a33a8e20fbcb0R192)
* optimized `vehicle::fuel_left` method by [caching all parts that can hold fuel](https://github.com/CleverRaven/Cataclysm-DDA/pull/41780/commits/786a5269c704dd54a261e4bf4e0ff803ba687fb3#diff-3d6b9eaf8fd915cd10858174d45c8d6aR3249)
   * this one is important, as it improved performance of all vehicular code (not only related smart controller). `vehicle::thrust` and `vehicle::idle` time improved twofold!
* improved performance of `vehicle::player_in_control`, [by changing order of if clauses](https://github.com/CleverRaven/Cataclysm-DDA/pull/41780/commits/786a5269c704dd54a261e4bf4e0ff803ba687fb3#diff-3d6b9eaf8fd915cd10858174d45c8d6aR273)
    * this one is not that important, because  `player_in_control` is not called as often as `fuel_left`


Profiling results:

1. Vehicle driving for 30 turns, speed is constantly changing to bust smart controller's cache:

Before optimization:
![image](https://user-images.githubusercontent.com/2865203/86542917-d982f480-bece-11ea-871c-eb4f198d48c5.png)

After optimization:
![image](https://user-images.githubusercontent.com/2865203/86542933-f28ba580-bece-11ea-8386-817ebda75b8a.png)

Night-Pryanik commented:
Messages about smart controller optimized engine state seems like too frequent. Does player ever need this information at all? I think this optimization could be done without any message at all (maybe only debug message) as player _expects_ optimization from installed smart controller.

Aivean commented:
> Messages about smart controller optimized engine state seems like too frequent. Does player ever need this information at all? I think this optimization could be done without any message at all (maybe only debug message) as player _expects_ optimization from installed smart controller.

@Night-Pryanik , are there any guidelines regarding log messages? Not arguing about that particular case, just for the future reference.

Aivean commented:
"Smart controller optimizes engine state." message is now visible only in debug mode.  
Currently smart controller prints log messages only when it's shutting down due to unsupported engine configuration or vehicle type.

mlangsdorf commented on src/vehicle.cpp:
```diff
@@ -3242,15 +3244,18 @@ point vehicle::pivot_displacement() const
-            return lhs;
+    int fl = 0;
+
+    for( size_t i = 0; i < fuel_containers.size(); ++i ) {
```
```suggestion
    for( const size_t i : fuel_containers ) {
        const vehicle_part &part = parts[i];
```
right?  You're just looping through fuel_containers to get the index value into parts.

Aivean replied:
Yes, there is this one and bunch of other issues that are failing clang-tidy build. I'll try to fix them all today.

Alex-Folts commented:
Do you aware that most fuel economy in hybrid vehicle comes from recuperative braking(in particular it is the case for early Toyota Prius) in city trafic? I had this idea recently about adding some recuperative braking controller, now i see your smart controller and think you must implement it.

The idea is that major price in energy you pay to accelerate vehicle(and i think in game it is work the same, but not sure), recuperative braking system use electric motor in reverse(as generator) mode to slow down vehicle and charge batteries returning some % of the energy moving vehicle had.

In game it should be something like this: if vehicle have electric motor(and battery, and recuperative braking controller), when player trying to decelerate vehicle you calculate how much kinetic energy will be dissipated(delta v times vehicle mass... etc), and send this energy into batteries respecting electric_to_kinetic_energy efficiency of said electric motor, but no more than max power of the electric motor.

You could implement this feature since you touch this subject already.

Aivean commented:
@Alex-Folts ,  I was not aware of the efficiency of regenerative braking. [It's fascinating](https://en.wikipedia.org/wiki/File:Energy_flow_in_cars_(recuperation_in_green).pdf). I'll look into implementing it.

However, I'm not sure that regenerative braking should be a feature of the "smart controller" or that it should require additional controller. Feels like this is something that can be inherent to the electric motors in CDDA. I don't think that this is something that players would want to disable or control, hence, no need for the external controller.

Alex-Folts commented:
> However, I'm not sure that regenerative braking should be a feature of the "smart controller" or that it should require additional controller. Feels like this is something that can be inherent to the electric motors in CDDA. I don't think that this is something that players would want to disable or control, hence, no need for the external controller.

Well usually it is part of the electric motor controller which is separate module, it is basicaly just battery charger adopted to specific use, and some controll hardware(i think this is your smart controller) which say when and hot it used. 
Since cdda have no electric motor controller part, you can put all functionality in your smart motor controller, or add new motor controller vehicle part which will have partial functionality of regerative breaking, but require also system to controll it(which included in your smart controller).
As for why separate module, well it is not required for vehicle to move, it is efficiency improvement, so you can find vehicle with broken part and vehicle still will work, you will need to fix it, learn some skills, search parts, or search cars hunting for working part - addition layer for gameplay.
